File: backend/tools/coverage_baseline.py
import json
import os
import sys
import xml.etree.ElementTree as ET


def _round_down_1(x: float) -> float:
    return float(int(x * 10)) / 10.0


def _pct(rate: str | None) -> float:
    if rate is None:
        return 0.0
    try:
        return float(rate) * 100.0
    except ValueError:
        return 0.0


def _read_json(path: str) -> dict:
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def _write_json(path: str, obj: dict) -> None:
    with open(path, "w", encoding="utf-8") as f:
        json.dump(obj, f, indent=2, ensure_ascii=False)
        f.write("\n")


def _load_packages_coverage(coverage_xml_path: str) -> dict[str, dict[str, float]]:
    tree = ET.parse(coverage_xml_path)
    root = tree.getroot()

    pkgs: dict[str, dict[str, float]] = {}
    for pkg in root.findall(".//packages/package"):
        name = pkg.attrib.get("name", "")
        if not name:
            continue
        pkgs[name] = {
            "lines": _round_down_1(_pct(pkg.attrib.get("line-rate"))),
            "branches": _round_down_1(_pct(pkg.attrib.get("branch-rate"))),
        }
    return pkgs


def _expected_from_baseline(baseline: dict[str, float], slack: dict[str, float], floor: dict[str, float]) -> dict[str, float]:
    return {
        "lines": _round_down_1(max(float(floor.get("lines", 0)), float(baseline["lines"]) - float(slack.get("lines", 0)))),
        "branches": _round_down_1(max(float(floor.get("branches", 0)), float(baseline["branches"]) - float(slack.get("branches", 0)))),
    }


def _best_match(package_name: str, critical_prefixes: list[str]) -> str | None:
    best = None
    for p in critical_prefixes:
        if package_name == p or package_name.startswith(p + "."):
            if best is None or len(p) > len(best):
                best = p
    return best


def cmd_update(frontend_root: str, coverage_xml_path: str) -> int:
    baseline_path = os.path.join(frontend_root, "coverage.baseline.json")
    baseline_file = _read_json(baseline_path)
    critical = list(baseline_file.get("criticalPackages", []))
    baseline_map = dict(baseline_file.get("baseline", {}))

    current = _load_packages_coverage(coverage_xml_path)

    # Store baseline for the critical prefixes themselves (if present),
    # otherwise store a combined (min) coverage across matched subpackages.
    aggregates: dict[str, list[dict[str, float]]] = {p: [] for p in critical}
    for pkg_name, cov in current.items():
        best = _best_match(pkg_name, critical)
        if best is None:
            continue
        aggregates[best].append(cov)

    for prefix, entries in aggregates.items():
        if not entries:
            continue
        baseline_map[prefix] = {
            "lines": min(e["lines"] for e in entries),
            "branches": min(e["branches"] for e in entries),
        }

    baseline_file["baseline"] = baseline_map
    _write_json(baseline_path, baseline_file)
    print(f"Updated baseline: {baseline_path}")
    return 0


def cmd_check(frontend_root: str, coverage_xml_path: str) -> int:
    baseline_path = os.path.join(frontend_root, "coverage.baseline.json")
    baseline_file = _read_json(baseline_path)

    slack = baseline_file.get("slack", {})
    floor = baseline_file.get("minFloor", {})
    critical = list(baseline_file.get("criticalPackages", []))
    baseline_map: dict = baseline_file.get("baseline", {}) or {}

    current = _load_packages_coverage(coverage_xml_path)

    failures: list[str] = []
    checked: list[tuple[str, dict[str, float], dict[str, float]]] = []

    aggregates: dict[str, list[dict[str, float]]] = {p: [] for p in critical}
    for pkg_name, cov in current.items():
        best = _best_match(pkg_name, critical)
        if best is None:
            continue
        aggregates[best].append(cov)

    for prefix, entries in aggregates.items():
        base = baseline_map.get(prefix)
        if base is None or not entries:
            continue
        cur = {
            "lines": min(e["lines"] for e in entries),
            "branches": min(e["branches"] for e in entries),
        }
        exp = _expected_from_baseline(base, slack, floor)
        checked.append((prefix, cur, exp))
        if cur["lines"] + 1e-9 < exp["lines"]:
            failures.append(f"{prefix}: lines {cur['lines']}% < {exp['lines']}% (baseline {base['lines']}%, slack {slack.get('lines', 0)}%)")
        if cur["branches"] + 1e-9 < exp["branches"]:
            failures.append(
                f"{prefix}: branches {cur['branches']}% < {exp['branches']}% (baseline {base['branches']}%, slack {slack.get('branches', 0)}%)"
            )

    if failures:
        print("Coverage regression detected for critical backend packages:")
        for f in failures:
            print(f"  - {f}")
        print("")
        print("Fix:")
        print("  - Improve tests for the affected package(s), or")
        print("  - If the drop is intended, update baseline:")
        print("      cd backend && python tools/coverage_baseline.py update coverage.xml")
        return 1

    if checked:
        print("Backend critical coverage baseline check OK:")
        for pkg_name, cur, exp in sorted(checked, key=lambda x: x[0]):
            print(f"  - {pkg_name}: lines {cur['lines']}% (min {exp['lines']}%), branches {cur['branches']}% (min {exp['branches']}%)")
    else:
        print("No backend baseline entries to check yet (baseline missing).")

    return 0


def main() -> int:
    if len(sys.argv) != 3:
        print("Usage: python tools/coverage_baseline.py <check|update> <coverage.xml>")
        return 2

    cmd = sys.argv[1]
    coverage_xml = sys.argv[2]

    if cmd not in {"check", "update"}:
        print("Usage: python tools/coverage_baseline.py <check|update> <coverage.xml>")
        return 2

    backend_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    if cmd == "update":
        return cmd_update(backend_root, coverage_xml)
    return cmd_check(backend_root, coverage_xml)


if __name__ == "__main__":
    raise SystemExit(main())

File: frontend/scripts/coverage-baseline.mjs
import fs from 'node:fs/promises'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const frontendRoot = path.resolve(__dirname, '..')
const baselinePath = path.join(frontendRoot, 'coverage.baseline.json')
const coverageSummaryPath = path.join(frontendRoot, 'coverage', 'coverage-summary.json')

function roundDown1(n) {
  return Math.floor(n * 10) / 10
}

function assertPct(x, label) {
  if (typeof x !== 'number' || Number.isNaN(x)) {
    throw new Error(`Invalid percent for ${label}: ${x}`)
  }
}

function normalizeEntry(v) {
  return {
    lines: roundDown1(v.lines?.pct ?? 0),
    statements: roundDown1(v.statements?.pct ?? 0),
    functions: roundDown1(v.functions?.pct ?? 0),
    branches: roundDown1(v.branches?.pct ?? 0),
  }
}

async function readJson(p) {
  const txt = await fs.readFile(p, 'utf-8')
  return JSON.parse(txt)
}

async function writeJson(p, obj) {
  const txt = JSON.stringify(obj, null, 2) + '\n'
  await fs.writeFile(p, txt, 'utf-8')
}

function computeExpected(current, slack, floor) {
  const expected = {}
  for (const k of ['lines', 'statements', 'functions', 'branches']) {
    const s = slack?.[k] ?? 0
    const f = floor?.[k] ?? 0
    expected[k] = roundDown1(Math.max(f, current[k] - s))
  }
  return expected
}

function formatRow(label, cur, exp) {
  return `${label}: L ${cur.lines}% (min ${exp.lines}%), S ${cur.statements}% (min ${exp.statements}%), F ${cur.functions}% (min ${exp.functions}%), B ${cur.branches}% (min ${exp.branches}%)`
}

async function cmdUpdate() {
  const baselineFile = await readJson(baselinePath)
  const summary = await readJson(coverageSummaryPath)

  const critical = baselineFile.critical ?? []
  const baseline = baselineFile.baseline ?? {}

  for (const key of critical) {
    const v = summary[key]
    if (!v) {
      // Keep entry absent if Jest didn't report it (e.g., excluded / not collected).
      continue
    }
    baseline[key] = normalizeEntry(v)
  }

  baselineFile.baseline = baseline
  await writeJson(baselinePath, baselineFile)
  console.log(`Updated baseline: ${path.relative(frontendRoot, baselinePath)}`)
}

async function cmdCheck() {
  const baselineFile = await readJson(baselinePath)
  const summary = await readJson(coverageSummaryPath)

  const slack = baselineFile.slack ?? {}
  const floor = baselineFile.minFloor ?? {}
  const critical = baselineFile.critical ?? []
  const baseline = baselineFile.baseline ?? {}

  const failures = []

  for (const key of critical) {
    const curRaw = summary[key]
    const baseRaw = baseline[key]

    if (!curRaw || !baseRaw) {
      // If there's no baseline yet, don't fail. Ask to update baseline.
      continue
    }

    const cur = normalizeEntry(curRaw)
    const exp = computeExpected(baseRaw, slack, floor)

    for (const metric of ['lines', 'statements', 'functions', 'branches']) {
      assertPct(cur[metric], `${key}.${metric}`)
      if (cur[metric] + 1e-9 < exp[metric]) {
        failures.push(`${key}: ${metric} ${cur[metric]}% < ${exp[metric]}% (baseline ${baseRaw[metric]}%, slack ${slack?.[metric] ?? 0}%)`)
      }
    }
  }

  if (failures.length) {
    console.error('Coverage regression detected for critical files:')
    for (const f of failures) console.error(`  - ${f}`)
    console.error('')
    console.error('Fix:')
    console.error('  - Improve tests for the affected module(s), or')
    console.error('  - If the drop is intended, update baseline:')
    console.error('      cd frontend && npm run coverage:baseline:update')
    process.exit(1)
  }

  // Helpful output for CI logs
  const checked = critical.filter((k) => summary[k] && baseline[k])
  if (checked.length) {
    console.log('Critical coverage baseline check OK:')
    for (const key of checked) {
      const cur = normalizeEntry(summary[key])
      const exp = computeExpected(baseline[key], slack, floor)
      console.log(`  - ${formatRow(key, cur, exp)}`)
    }
  } else {
    console.log('No critical baseline entries to check yet (baseline missing).')
  }
}

const cmd = process.argv[2]
if (!cmd || !['check', 'update'].includes(cmd)) {
  console.error('Usage: node scripts/coverage-baseline.mjs <check|update>')
  process.exit(2)
}

try {
  if (cmd === 'update') await cmdUpdate()
  else await cmdCheck()
} catch (e) {
  console.error(e?.stack || String(e))
  process.exit(1)
}

